multilep/src/GenAnalyzer.cc

```diff
@@ -105,8 +125,84 @@ void GenAnalyzer::analyze(const edm::Event& iEvent){
                 _gen_phPassParentage[_gen_nPh] = GenTools::passParentage(p, *genParticles);
                 ++_gen_nPh;
             } 
-        }
-    }
+	}
+
+       //store all generator level particles
+       if( multilepAnalyzer->storeGenParticles )
+	 {	    
+	    float ptGen = p.pt();
+	    float etaGen = p.eta();
+	    float phiGen = p.phi();
+	    float EGen = p.energy();
+	    int idGen = p.pdgId();
+	    int statusGen = p.status();
+	    int chargeGen = p.charge();
+	    bool isPromptFinalStateGen = p.isPromptFinalState();
+	    bool isDirectPromptTauDecayProductFinalState = p.isDirectPromptTauDecayProductFinalState();
+	    bool isLastCopy = p.isLastCopy();
+	    int indexGen = _gen_n;
+
+	    if( _gen_n == gen_n_max )
+	      {		      
+		 std::cout << "Reached max" << std::endl;
+		 exit(1);
+	      }
+       
+	    const reco::GenParticle* mom = GenTools::getMother(p, *genParticles);
+       
+	    int mother_index = 0;
+	    if( mom )
+	      {		 
+		 reco::GenParticleCollection genParticles_m = *genParticles;
+		 for(reco::GenParticleCollection::const_iterator pp = genParticles_m.begin(); pp != genParticles_m.end(); pp++)
+		   {
+		      if( fabs(pp->pt()-mom->pt()) < 10E-6 && fabs(pp->eta()-mom->eta()) < 10E-6 ) break;
+		      mother_index++;
+		   }
+	      }
+	    else mother_index = -1;
+       
+	    _gen_daughter_n[_gen_n] = 0;
+       
+	    const reco::GenParticleRefVector& daughterRefs = p.daughterRefVector();
+       
+	    for(reco::GenParticleRefVector::const_iterator idr = daughterRefs.begin(); idr!= daughterRefs.end(); ++idr)
+	      {	    
+		 if( idr->isAvailable() )
+		   {
+		      const reco::GenParticleRef& genParticle = (*idr);
+		      const reco::GenParticle *d = genParticle.get();
+		      
+		      int index = 0;
+		      reco::GenParticleCollection genParticles_d = *genParticles;
+		      for(reco::GenParticleCollection::const_iterator pp = genParticles_d.begin(); pp != genParticles_d.end(); pp++)
+			{
+			   reco::GenParticle *mcp_d = &(const_cast<reco::GenParticle&>(*pp));
+			   if( fabs(mcp_d->pt()-(*d).pt()) < 10E-6 && fabs(mcp_d->eta()-(*d).eta()) < 10E-6 ) break;
+			   index++;
+			}
```

Comment:
I would think here you might be able to simplify in a similar way by making use of index=d.daughterRef(idr).key() with idr simply the integer in range 0 to ndaughters
(of course I might be wrong, so good to double check if those really give the same)